### DISCUSSION TITLE
Logging every command output

### DISCUSSION THREAD

**Discussion by b1two** [kovidgoyal/kitty#5713] [Q&A] [answered]
Hello everyone,

I'd like to be able to log every command output automatically to files. This would be helpful so that one does not need to take note of everything and yet be able to retrieve any piece of information. In the end, a filter may be implemented to limit which command is logged, to what file you save the output, and so on.

I first though of using a ZSH plugin, however it does not seem that simple, since I do not want to mess with any redirection and all from the command line.

So I decided to look into building a new kitten. I figured by using the _output_ type of input to the kitten I am able to retrieve the content of the last command and then do whatever I want with it. However I did not manage to retrieve the command line from the kitten (the command that was actually run). Do you know any solution to also retrieve the last command info? (from the shell extension probably)

Finally, I would love to have the kitten run each time a command finishes, and not just when I press a shortcut. Do you think it is possible? Would I have to enable the remote control of Kitty? I am kind of reluctant to use such thing as it would increase the attack surface of Kitty.

Thank you, and thank you kovidgoyal and al. for this great tool!

**Comment by kovidgoyal** [comment, marked as answer]
I don't understand your motivation for this, you can simply use the
existing scrollback, which you can save to a file at anytime you like by
pressing ctrl+shift+h and then using the save command in less.

However if you really want to save commands and outputs, it's not that
hard, you just need to save from the first prompt line before the last
output start. See for example how dump_lines_with_attrs works. Prompt
start and output start are attrs. You can send a PR to add a function to
get the last prompt just like there is one to the get last output.

As for doing it automatically, you will need to use remote control. But
given that remote control has capability based security you can lock it
down as tightly as you need.

**Comment by b1two** [comment]
Thank you for the response.

The main motivation is that during security assessments (Red Teaming,
internal pentesting or even external pentesting), it is often required to
provide exact timings of when the command was executed, what exactly
happened, …  (sometimes several days/weeks after the end of the
assessment). In the same context, it can be useful to have everything
you did when you are writing the report: as some assessment last for
several weeks, you probably already have closed your terminal several
times and are not able to get traces back that you first though would be
useless (and thus traces that you did not save at the time).

Thanks for the pointer, it has been a long time since I have touched C,
and never used python bindings in C. I will probably take a look in the
future, however in the mean time I used a (crappy) workaround, directly
parsing the _dump_lines_with_attrs_ output.

The kitten code is available here, if anyone is interested:
https://gist.github.com/b1two/5d5fb26d086c8df989d52cfacef0432a

I will then take a look at the automation via remote control.

Thank you.

**Comment by kovidgoyal** [comment]
Isn't the script command designed for exactly this? It even has
scriptreplay. Anyway glad you got something that works for you.

**Comment by b1two** [comment]
Yes, I dug a little into script but it handles poorly the TUI applications
such as vim when reading the plaintext log file (without using the
scriptreplay feature). So I was looking to profit from all the handling
and formatting of Kitty to do so.
I'll mark your first message as answer, and will post the final script
here once it's clean.

**Reply by hainkind** [reply]
@b1two i'm also interested in this, any success with your script?

**Comment by may** [comment]
I've been looking for this recently, and the best option I've come up with after trying several is to simply use 'script' as a backup, in case I forget to save it via `less`.

The following dumps a text file in the current directory with a filename of the current time, and you can see everything you typed and the program output (logged all stdin & stdout).

`script $(date '+%Y-%m-%dT%H-%M-%S').txt`

Should be built in on most Linux machines and maybe MacOS.

Ideally the tool I'm using would do the logging itself, but this is sufficiently useful that it gets me where I want to go.

**Comment by xuyungit** [comment]
I recently realized that I might need this feature as well: sometimes the terminal suddenly loses its history, and the mouse scrollback only goes through the command history, not the buffer. In this situation, I can hardly find any way to retrieve the historical output information.

**Comment by Kimononono** [comment]
I revised @b1two 's code into a watcher that can log { command, output } to wherever whenever a command finishes running.

Kitty having inline labels like `\x1b[32mprompt \x1b[39m`, `\x1b[33moutput \x1b[39m` makes this (hopefully) alot more stable than my prior implementation using script. Also TUI apps dont spam the log :) 

https://gist.github.com/Kimononono/aca06eb1885e485542a5be09f19bc30c
